connectors/src/connectors/notion/lib/parents.ts

```diff
@@ -0,0 +1,157 @@
+import {
+  DataSourceConfig,
+  DataSourceInfo,
+} from "@connectors/types/data_source_config";
+
+import {
+  getDatabaseChildrenOfDocument,
+  getNotionPageFromConnectorsDb,
+  getPageChildrenOfDocument,
+} from "./connectors_db_helpers";
+
+/** Compute the parents field for a notion document See the [Design
+ * Doc](https://www.notion.so/dust-tt/Engineering-e0f834b5be5a43569baaf76e9c41adf2?p=3d26536a4e0a464eae0c3f8f27a7af97&pm=s)
+ * and the field documentation [in
+ * core](https://github.com/dust-tt/dust/blob/main/core/src/data_sources/data_source.rs)
+ * for relevant details
+ *
+ * @param memoizationKey optional key to control memoization of this function (not actually used by the functio)
+ *
+ */
+async function _getParents(
+  dataSourceInfo: DataSourceInfo,
+  document: {
+    notionId: string;
+    parentType: string | null | undefined;
+    parentId: string | null | undefined;
+  },
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars -- used for memoization
+  memoizationKey?: string
+): Promise<string[]> {
+  const parents: string[] = [document.notionId];
+  switch (document.parentType) {
+    case "workspace":
+      return parents;
+    case "block":
+      // rare cases in which doing something here is useful
+      // are ignored for now, see the design doc for details
+      return parents;
+    case "page":
+    case "database": {
+      // retrieve the parent from notion connectors db
+      // and add it to the parents array
+      const parent = await getNotionPageFromConnectorsDb(
+        dataSourceInfo,
+        document.parentId as string // (cannot be null here)
+      );
+      if (!parent) {
+        // The parent is either not synced yet (not an issue, see design doc) or
+        // is not in Dust's scope, in both cases we can just return the page id
+        return parents;
+      }
+      return parents.concat(
+        await getParents(dataSourceInfo, {
+          notionId: parent.notionPageId,
+          parentType: parent.parentType,
+          parentId: parent.parentId,
+        })
+      );
+    }
+    default:
+      throw new Error(`Unhandled parent type ${document.parentType}`);
+  }
+}
+
+export const getParents = memoize(
+  _getParents,
+  (dataSourceInfo, document, memoizationKey) => {
+    return `${dataSourceInfo.dataSourceName}:${document.notionId}:${memoizationKey}`;
+  }
+);
+
+export async function updateAllParentsFields(
+  dataSourceConfig: DataSourceConfig,
+  documents: (NotionPage | NotionDatabase)[]
+) {
+  /* Computing all descendants, then updating, ensures the field is updated only
+    once per page, limiting the load on the Datasource */
+  const pagesToUpdate = await getPagesToUpdate(documents, dataSourceConfig);
+
+  // Update everybody's parents field. Use of a memoization key to avoid
+  // potentially sharing memoization across updateAllParentsFields calls, which
+  // would be incorrect
+  const memoizationKey = uuid4();
```

Comment:
If not too hard, let's pass the activityId (examples in Slack connector I believe) instead of using a uuid as it will be more deterministic and will be reused if the activity is restarted, eg due to too many Notion rate limit errors as an example

Reply:
Or the same as the other if it makes sense, but something stable-ish?